### PR TITLE
Add `-Raw` Experimental switch to `ConvertTo-Json` cmdlet

### LINKED ISSUE #5797: ConvertTo-Json: unexpected behavior with objects that have ETS properties (NoteProperty, ScriptProperty)
**Instances of types that normally serialize to a JSON _scalar_ (basic JSON data type: string, number, Boolean) should not serialize to _objects_ just because  - situationally - NoteProperty and ScriptProperty members may be present - not least because such members may be added _automatically_ by PowerShell.**

Additionally, the current behavior is _inconsistent_, possibly related to #5579.

Also, with a `ScriptProperty` member present on a `[string]` instance, `ConvertTo-Json` _crashes_ as of PowerShell Core v6.1.0-preview.3 - see #7091

Steps to reproduce
------------------

```powershell
[ordered] @{ d1 = [datetime]::now; d2 = get-date  } | ConvertTo-Json
```

Expected behavior
-----------------

```none
{
  "d1": "2018-01-05T09:25:37.037783+01:00",
  "d2": "2018-01-05T09:25:37.037783+01:00",
}
```

Actual behavior
---------------

```none
{
  "d1": "2018-01-05T09:25:37.037783+01:00",
  "d2": {
    "value": "2018-01-05T09:25:37.037915+01:00",
    "DisplayHint": 2,
    "DateTime": "Friday, January 5, 2018 9:25:37 AM"
  }
}
```

Note how the presence of  the `DisplayHint` and `DateTime` members added by `Get-Date` caused the value to no longer serialize as a single string, but as an _object_ with said members (yet not any of the type's regular properties).

In some cases, using an intermediate variable makes the problem go away:

```powershell
$dt = get-date
[ordered] @{ d1 = [datetime]::now; d2 = $dt  } | ConvertTo-Json # OK
```

That said, with other objects, such as those with provider-added properties, the problem surfaces even with an intermediate variable:

```powershell
'hi' > t.txt
$s = Get-Content t.txt
$s | ConvertTo-Json   # !! outputs a JSON object with many provider properties
```

Again, what normally serializes as a _string_ is unexpectedly serialized as an _object_  due to the presence of NoteProperty members.

Finally, when sending an instance with even just a ScriptProperty member _through the pipeline_, the problem surfaces too:

```powershell
> [datetime]::now | ConvertTo-Json
{
  "value": "2018-01-07T06:18:29.821274+01:00",
  "DateTime": "Sunday, January 7, 2018 6:18:29 AM"
}
```

Note that if the instance is nested inside a hashtable / custom object, the problem does _not_ occur, as demonstrated above.
It also doesn't occur if you use `-InputObject` instead of the pipeline: `ConvertTo-Json  ([datetime]::now)`

Workarounds
-------------

Either: apply `.psobject.baseobject` to the command producing the object to serialize to JSON:

```powershell
@{ d = (Get-Date).psobject.BaseObject } | ConvertTo-Json
```

Or: cast to the expected type (or, with an intermediate variable, type-constrain it):

```powershell
@{ d = [datetime] (Get-Date) } | ConvertTo-Json
```

Note: The above workarounds only work if the problematic object is not the input object as a whole.

If so, the only workaround I'm aware of is to pass the input object via `-InputObject` rather than the pipeline:

```powershell
ConvertTo-Json ([datetime] (Get-Date))   # OK, due to using -InputObject

# !! BROKEN, due to using the pipeline, though, curiously, only
# !! the "DateTime" ScriptProperty shows up, not the "DisplayHint" NoteProperty
[datetime] (Get-Date) | ConvertTo-Json  
```

Environment data
----------------

```powershell
PowerShell Core v6.0.0-rc.2 (v6.0.0-rc.2) on macOS 10.13.2
```

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

Objects that contain ETS members can cause problems when serializing to JSON.  In the case of script properties on a string, it gets stuck in a recursive loop.

As discussed in https://github.com/PowerShell/PowerShell/issues/5797, proposal is to have an ExperimentalFeature to add a `-Raw` switch that will use the BaseObject of the PSObject (effectively removing ETS members) for serialization.

## PR Context

Fix https://github.com/PowerShell/PowerShell/issues/5797

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [x] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [x] Experimental feature name(s): Microsoft.PowerShell.Utility.PSConvertToJsonRaw
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/6847
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by iSazonov** [comment]
Although this seems to solve this particular problem, it is doubtful. The appearance of such a Raw parameter indicates that we have a poor understanding of whether we should serialize ETS properties or not. If these properties are required, then why should we exclude them from serialization? If so, is it always? If always and if we are talking about serialization, then _attributes_ are usually used to exclude properties from serialization. This approach is more general and does not require new parameters in cmdlets. If not always, then perhaps we need something like Select-Object -NoETSProperties or new cmdlet Get-PSBaseObject.

**Comment by SteveL-MSFT** [comment]
@iSazonov I would suggest having this conversation in the issue.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertToJsonCommand.cs [hunk 1]
```diff
@@ -100,7 +109,14 @@ protected override void BeginProcessing()
         /// </summary>
         protected override void ProcessRecord()
         {
-            _inputObjects.Add(InputObject);
+            if (ExperimentalFeature.IsEnabled("Microsoft.PowerShell.Utility.PSConvertToJsonRaw") && Raw && InputObject is PSObject)
```
Comment (TylerLeonhardt) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertToJsonCommand.cs:112]:
```suggestion
            if (ExperimentalFeature.IsEnabled("Microsoft.PowerShell.Utility.PSConvertToJsonRaw")
                && Raw
                && InputObject is PSObject obj)
```

this along with using `obj` below.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertToJsonCommand.cs [hunk 2]
```diff
@@ -100,7 +109,14 @@ protected override void BeginProcessing()
...
+            {
+                _inputObjects.Add(((PSObject)InputObject).BaseObject);
```
Comment (TylerLeonhardt) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertToJsonCommand.cs:114]:
```suggestion
                _inputObjects.Add(obj.BaseObject);
```
this along with declaring `obj` above

**Review by TylerLeonhardt — COMMENTED** [review]
less parens

**Review by TylerLeonhardt — APPROVED** [review]
tbh I don't really like `-Raw`... it's not clear by the name... but I can't think of anything better right now so LGTM.

Really this is "stripping off ETS properties" but `-StripETSProps` `-StripNoteProperties` or anything like that seems overly verbose.

`-UseBaseObject`? Idk.

**Comment by TylerLeonhardt** [comment]
I almost feel like `Select-Object` should _really_ have the `-Raw` parameter so you can do:

```
$obj | Select-Object -Raw | ConvertTo-Json
$obj | Select-Object -Raw | ConvertTo-Csv
($obj | Select-Object -Raw
$obj | Select-Object -Raw | ConvertTo-Yaml
```

**Comment by TylerLeonhardt** [comment]
Or going down the route of having a `Remove-Member` or something like that that will remove a specific ETS property or _all_ ETS properties.

**Comment by iSazonov** [comment]
@TylerLeonhardt We continue the discussion in related issue. I believe we should investigate the issue more in depth.

**Comment by mklement0** [comment]
@TylerLeonhardt:

As for why `Select-Object -Raw` isn't an option: see https://github.com/PowerShell/PowerShell/issues/5797#issuecomment-723134446

As for why `-Raw` isn't a great name, but is the closest thing to a _pattern_ that we have, after `-Bare` was turned down in order to communicate the desire to output an _undecorated / unwrapped_ object: see #7855

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertToJsonCommand.cs (hunk 1 above)
Comment (daxian-dbw) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertToJsonCommand.cs:112]: The call to `ExperimentalFeature.IsEnabled("Microsoft.PowerShell.Utility.PSConvertToJsonRaw")` is not needed here. `Raw` is guarded by the experimental attribute, so when the fetaure is disabled, a user cannot see `-Raw` and thus `Raw` will always be false.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertToJsonCommand.cs (hunk 2 above)
Comment (daxian-dbw) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertToJsonCommand.cs:114]: What if `obj` is a pure PSObject that is not backed up by a .NET object? I guess we should write out error in that case?

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee decided on a different approach https://github.com/PowerShell/PowerShell/issues/5797#issuecomment-747107494

**Pull request closed without merging**